fvm/src/syscalls/crypto.rs

```diff
@@ -4,46 +4,76 @@ use std::cmp;
 
 use anyhow::{anyhow, Context as _};
 use fvm_shared::crypto::signature::{
-    SignatureType, SECP_PUB_LEN, SECP_SIG_LEN, SECP_SIG_MESSAGE_HASH_SIZE,
+    BLS_DIGEST_LEN, BLS_PUB_LEN, BLS_SIG_LEN, SECP_PUB_LEN, SECP_SIG_LEN,
+    SECP_SIG_MESSAGE_HASH_SIZE,
 };
 use fvm_shared::piece::PieceInfo;
 use fvm_shared::sector::{
     AggregateSealVerifyProofAndInfos, RegisteredSealProof, ReplicaUpdateInfo, SealVerifyInfo,
     WindowPoStVerifyInfo,
 };
 use fvm_shared::sys;
-use num_traits::FromPrimitive;
 
 use super::Context;
 use crate::kernel::{ClassifyResult, Result};
 use crate::{syscall_error, Kernel};
 
-/// Verifies that a signature is valid for an address and plaintext.
+/// Verifies that a bls aggregate signature is valid for a list of public keys and plaintext
+/// digests.
 ///
 /// The return i32 indicates the status code of the verification:
 ///  - 0: verification ok.
 ///  - -1: verification failed.
-#[allow(clippy::too_many_arguments)]
-pub fn verify_signature(
+pub fn verify_bls_aggregate(
     context: Context<'_, impl Kernel>,
-    sig_type: u32,
+    num_signers: u32,
     sig_off: u32,
-    sig_len: u32,
-    addr_off: u32,
-    addr_len: u32,
-    plaintext_off: u32,
-    plaintext_len: u32,
+    pub_keys_off: u32,
+    digests_off: u32,
 ) -> Result<i32> {
-    let sig_type = SignatureType::from_u32(sig_type)
-        .with_context(|| format!("unknown signature type {}", sig_type))
-        .or_illegal_argument()?;
-    let sig_bytes = context.memory.try_slice(sig_off, sig_len)?;
-    let addr = context.memory.read_address(addr_off, addr_len)?;
-    let plaintext = context.memory.try_slice(plaintext_off, plaintext_len)?;
+    let pub_key_ref_len = std::mem::size_of::<&[u8; BLS_PUB_LEN]>();
+    let digest_ref_len = std::mem::size_of::<&[u8; BLS_DIGEST_LEN]>();
+
+    // Check that the provided number of signatures aggregated does not cause `u32` overflow.
+    let (pub_key_refs_len, digest_refs_len) = num_signers
+        .checked_mul(pub_key_ref_len as u32)
+        .zip(num_signers.checked_mul(digest_ref_len as u32))
+        .ok_or(syscall_error!(
+            IllegalArgument;
+            "number of signatures aggregated ({}) exceeds limit",
+            num_signers
+        ))?;
```

Comment:
Ah, sorry. I meant:

1. Keep the reference to a slice of fixed-size arrays types.
2. Avoid copying those arrays.

I.e.:

```rust 
let pub_keys: &[[u8; BLS_PUB_LEN]] = unsafe {
    // (checking for overflows)
    let buf = context.memory.try_slice(pub_keys_off, num_signers * BLS_PUB_LEN)?;
    std::slice::from_raw_parts(pub_keys.as_ptr(), num_signers)
};
```

(we can also debug/static assert the layout/alignment, but that shouldn't be necessary)

Reply:
We could also make this "nicer" by adding a helper to the `Memory` type. I.e.:

```rust
fn try_slice_chunks<'a, const S: usize>(&'a self, offset: u32, len: u32) -> &'a [[u8; S]];
```

Reply:
I changed the keys/digests parameters from slices of fixed-size arrays, e.g. `&[[u8; BLS_DIGEST_LEN]]`, to slices of byte slices `&[&[u8]]` because the SDK's `verify_signature` needs to operate on digests/public keys of variable lengths depending on BLS vs Secp signatures. So, in order to to pass the variable-size digests/public keys from the SDK to the Kernel as fixed-size arrays, I would need to perform copies at the SDK layer.

I can revert this change so that the `verify_bls_aggregate` syscall expects `&[[u8; SOME_LEN]]`, however I do think copies at the SDK layer would be required (I'm not sure if that's acceptable or not). If that's ok, then I can add the memory chunking helper.

Reply:
Copies in the SDK are fine, copies in the kernel/FVM are less great. Basically, we have to explicitly charge for everything we do in the FVM itself, so I prefer to do things in "userland" (where the SDK lives in userland).

Reply:
It also lets the user (in a future world where we allow arbitrary wasm actors) optimize this operation to avoid the copy, if they care about it. But once we pass from the SDK and into the syscall, the user can't change anything.

Reply:
Cool, thanks. That's good info